### PR TITLE
feat(api): DNS challenge support for Let's Encrypt

### PR DESCRIPTION
**Pull request by nils-wisiol**
Provides easy setting of TXT records at `_acme-challenge.yourname.dedyn.io.`

A certbot hook script can be found at https://github.com/desec-io/certbot-hook

### CODE REVIEW

File: api/desecapi/migrations/0013_acme_challenge.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+    dependencies = [
+        ('desecapi', '0012_move_dyn_flag'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='domain',
+            name='acme_challenge',
+            field=models.CharField(max_length=1024, blank=True),
```
Comment (peterthomassen) [inline api/desecapi/migrations/0013_acme_challenge.py]: The length of a string in a `TXT` RR is limited to 255, so this field is too wide. I leave if up to your judgment to reflect that fact here or not.

File: api/desecapi/models.py [hunk 2]
```diff
@@ -99,6 +99,7 @@ class Domain(models.Model):
     arecord = models.CharField(max_length=255, blank=True)
     aaaarecord = models.CharField(max_length=1024, blank=True)
     owner = models.ForeignKey(settings.AUTH_USER_MODEL, related_name='domains')
+    acme_challenge = models.CharField(max_length=1024, blank=True)
```
Comment (peterthomassen) [inline api/desecapi/models.py]: See above

File: api/desecapi/tests/testdomains.py [hunk 3]
```diff
@@ -189,13 +199,29 @@ def testPostingWithRecordsCausesPdnsAPIPatch(self):
...
+        data = {'name': name, 'arecord': '1.3.3.7', 'aaaarecord': 'dead::beef', 'acme_challenge': 'letsencrypt_ftw'}
         response = self.client.post(url, data)
 
         self.assertEqual(httpretty.last_request().method, 'PATCH')
         self.assertTrue(data['name'] in httpretty.last_request().parsed_body)
         self.assertTrue('1.3.3.7' in httpretty.last_request().parsed_body)
         self.assertTrue('dead::beef' in httpretty.last_request().parsed_body)
+        self.assertTrue('letsencrypt_ftw' in httpretty.last_request().parsed_body)
+
+    def testPatchDomainCausesPdnsAPIPatch(self):
```
Comment (peterthomassen) [inline api/desecapi/tests/testdomains.py]: This is a misnomer (you are posting to the domain, not patching)

File: api/desecapi/migrations/0013_acme_challenge.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+    dependencies = [
+        ('desecapi', '0012_move_dyn_flag'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='domain',
+            name='acme_challenge',
+            field=models.CharField(max_length=255, blank=True),
```
Comment (nils-wisiol) [inline api/desecapi/migrations/0013_acme_challenge.py]: length set to 255, however it is unclear if bytes or characters are counted (UTF-8)

File: api/desecapi/models.py [hunk 5]
```diff
@@ -99,6 +99,7 @@ class Domain(models.Model):
     arecord = models.CharField(max_length=255, blank=True)
     aaaarecord = models.CharField(max_length=1024, blank=True)
     owner = models.ForeignKey(settings.AUTH_USER_MODEL, related_name='domains')
+    acme_challenge = models.CharField(max_length=255, blank=True)
```
Comment (nils-wisiol) [inline api/desecapi/models.py]: updating any db that was build with the previous version of this PR will fail, drop database then rebuild from scratch (or change manually).

File: api/desecapi/tests/testdomains.py [hunk 6]
```diff
@@ -189,13 +199,29 @@ def testPostingWithRecordsCausesPdnsAPIPatch(self):
...
+        data = {'name': name, 'arecord': '1.3.3.7', 'aaaarecord': 'dead::beef', 'acme_challenge': 'letsencrypt_ftw'}
         response = self.client.post(url, data)
 
         self.assertEqual(httpretty.last_request().method, 'PATCH')
         self.assertTrue(data['name'] in httpretty.last_request().parsed_body)
         self.assertTrue('1.3.3.7' in httpretty.last_request().parsed_body)
         self.assertTrue('dead::beef' in httpretty.last_request().parsed_body)
+        self.assertTrue('letsencrypt_ftw' in httpretty.last_request().parsed_body)
+
+    def testPostDomainCausesPdnsAPIPatch(self):
```
Comment (nils-wisiol) [inline api/desecapi/tests/testdomains.py]: notice `post`

**Comment by nils-wisiol** [comment]
I addressed all requested changes; in an effort to streamline the review process I also added comments on the sections in the code that I changed.

I also first pushed with fixup commits, then rebased and squashed them and push --force again. (When trying how GitHub would deal with the outdated diffs in that case -- the handling did not improve.) That's how there is several commits showing on this page; but only one commit showing in the PR itself.

**Pull request merged**